internal/statemachine/common_states.go

```diff
@@ -138,29 +139,72 @@ func (stateMachine *StateMachine) generateDiskInfo() error {
 // on a 100MiB filesystem, ext4 takes a little over 7MiB for the
 // metadata. Use 8MB as a minimum padding here
 func (stateMachine *StateMachine) calculateRootfsSize() error {
-	rootfsSize, err := helper.Du(stateMachine.tempDirs.rootfs)
-	if err != nil {
-		return fmt.Errorf("Error getting rootfs size: %s", err.Error())
-	}
-	var rootfsQuantity quantity.Size = rootfsSize
-	rootfsPadding := 8 * quantity.SizeMiB
+	if stateMachine.commonFlags.Size != "" {
+		var parsedSize quantity.Size
+
+		// identify which structure has the rootfs
+		var rootfsVolume *gadget.Volume
+		var rootfsVolumeName string
+		for volumeName, volume := range stateMachine.GadgetInfo.Volumes {
+			for _, structure := range volume.Structure {
+				if structure.Size == 0 {
+					rootfsVolume = volume
+					rootfsVolumeName = volumeName
+				}
+			}
+		}
 
-	// fudge factor for incidentals
-	rootfsQuantity = quantity.Size(math.Ceil(float64(rootfsQuantity) * 1.5))
-	rootfsQuantity += rootfsPadding
+		if !strings.Contains(stateMachine.commonFlags.Size, ":") {
+			// this scenario has just one size for each volume
+			// no need to check error as it has already been done by
+			// the parseImageSizes function
+			parsedSize, _ = quantity.ParseSize(stateMachine.commonFlags.Size)
+		} else {
+			parsedSize = stateMachine.ImageSizes[rootfsVolumeName]
+		}
 
-	// align the size of the rootfs to sector size
-	rootfsQuantity = quantity.Size(math.Ceil(float64(rootfsQuantity)/float64(stateMachine.SectorSize))) *
-		quantity.Size(stateMachine.SectorSize)
+		// subtract the size and offsets of the existing volumes
+		if rootfsVolume != nil {
+			for _, structure := range rootfsVolume.Structure {
+				parsedSize -= structure.Size
+				if structure.Offset != nil {
+					parsedSize -= quantity.Size(*structure.Offset)
+				}
+			}
+		}
 
-	stateMachine.RootfsSize = rootfsQuantity
+		// align the size of the rootfs to sector size
+		parsedSize = quantity.Size(math.Ceil(float64(parsedSize)/float64(stateMachine.SectorSize))) *
+			quantity.Size(stateMachine.SectorSize)
+
+		stateMachine.RootfsSize = parsedSize
```

Comment:
In this case, should we still check if the newly calculated size can fit the current rootfs? So maybe calculating the size with `du` regardless of whether the image size is given and (preferred) erroring out when there's not enough space? Or is there some check for that later?

Reply:
Yes, good point. I don't think there is a check for this later. I'll add logic to always do the du.